### PR TITLE
Address concerns with storagenode Retain endpoint

### PR DESCRIPTION
**Pull request by mobyvb**
What: 
PR #2424 was merged and it turns out there are a few things we needed to fix. We fixed them in #2366, but we decided to split them out into a separate PR to help reviewers.

Review and merge before #2366 because it will ~halve the changed lines/files in that pr!

Why:

Please describe the tests:
 - Test 1:
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

File: storagenode/piecestore/endpoint.go [hunk 1]
```diff
@@ -547,25 +548,37 @@ func (endpoint *Endpoint) SaveOrder(ctx context.Context, limit *pb.OrderLimit, o
 func (endpoint *Endpoint) Retain(ctx context.Context, retainReq *pb.RetainRequest) (*pb.RetainResponse, error) {
 	peer, err := identity.PeerIdentityFromContext(ctx)
 	if err != nil {
-		return nil, Error.Wrap(err)
+		return nil, status.Error(codes.Internal, Error.Wrap(err).Error())
+	}
+
+	trustedSatellites := endpoint.trust.GetSatellites(ctx)
```
Comment (egonelbre) [inline storagenode/piecestore/endpoint.go]: Add a new method rather than copying all the information.

File: storagenode/piecestore/endpoint.go [hunk 2]
```diff
@@ -547,25 +548,37 @@ func (endpoint *Endpoint) SaveOrder(ctx context.Context, limit *pb.OrderLimit, o
 func (endpoint *Endpoint) Retain(ctx context.Context, retainReq *pb.RetainRequest) (*pb.RetainResponse, error) {
 	peer, err := identity.PeerIdentityFromContext(ctx)
 	if err != nil {
-		return nil, Error.Wrap(err)
+		return nil, status.Error(codes.Internal, Error.Wrap(err).Error())
```
Comment (egonelbre) [inline storagenode/piecestore/endpoint.go]: this is not internal

File: storagenode/piecestore/endpoint.go [hunk 3]
```diff
@@ -547,25 +548,37 @@ func (endpoint *Endpoint) SaveOrder(ctx context.Context, limit *pb.OrderLimit, o
...
+		}
+	}
+	if !trusted {
+		return nil, status.Error(codes.PermissionDenied, Error.New("retain called with untrusted ID").Error())
 	}
 
 	filter, err := bloomfilter.NewFromBytes(retainReq.GetFilter())
 	if err != nil {
-		return nil, Error.Wrap(err)
+		return nil, status.Error(codes.Internal, Error.Wrap(err).Error())
```
Comment (egonelbre) [inline storagenode/piecestore/endpoint.go]: this is not internal error

File: storagenode/piecestore/endpoint.go (hunk 2 above)
Comment (navillasa) [reply storagenode/piecestore/endpoint.go]: what should it be instead then?

File: storagenode/piecestore/endpoint.go (hunk 3 above)
Comment (navillasa) [reply storagenode/piecestore/endpoint.go]: what should it be instead then?

File: storagenode/piecestore/endpoint.go (hunk 2 above)
Comment (egonelbre) [reply storagenode/piecestore/endpoint.go]: https://github.com/grpc/grpc/blob/master/doc/statuscodes.md

File: storagenode/piecestore/endpoint.go [hunk 4]
```diff
@@ -578,17 +586,30 @@ func (endpoint *Endpoint) Retain(ctx context.Context, retainReq *pb.RetainReques
...
 		offset += len(pieceIDs)
+		offset -= numDeleted
 		// We call Gosched() here because the GC process is expected to be long and we want to keep it at low priority,
 		// so other goroutines can continue serving requests.
 		runtime.Gosched()
 	}
 	return &pb.RetainResponse{}, nil
 }
 
+func (endpoint *Endpoint) isTrustedSatellite(ctx context.Context, peer *identity.PeerIdentity) bool {
```
Comment (egonelbre) [inline storagenode/piecestore/endpoint.go]: this should be in the pool implementation

Comment (navillasa) [reply storagenode/piecestore/endpoint.go]: that is, a method on `trust` instead of a method on `endpoint`?

Comment (egonelbre) [reply storagenode/piecestore/endpoint.go]: yes

Comment (navillasa) [reply storagenode/piecestore/endpoint.go]: okay changed it!

File: storagenode/trust/service.go [hunk 5]
```diff
@@ -154,3 +154,15 @@ func (pool *Pool) GetAddress(ctx context.Context, id storj.NodeID) (_ string, er
 	}
 	return info.nodeURL.Address, nil
 }
+
+// IsTrustedSatellite tells if a satellite is trusted or not
+func (pool *Pool) IsTrustedSatellite(ctx context.Context, peer *identity.PeerIdentity) bool {
+	trustedSatellites := pool.GetSatellites(ctx)
```
Comment (egonelbre) [inline storagenode/trust/service.go]: Do not copy the data

Comment (navillasa) [reply storagenode/trust/service.go]: ah okay sorry didn't understand before!

**Pull request merged**